jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/FragmentingFlusher.java

```diff
@@ -0,0 +1,109 @@
+//
+
+package org.eclipse.jetty.websocket.core.internal;
+
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.core.FrameHandler.Configuration;
+import org.eclipse.jetty.websocket.core.OpCode;
+
+/**
+ * Used to split large data frames into multiple frames below the maxFrameSize.
+ * Control frames and dataFrames smaller than the maxFrameSize will be forwarded
+ * directly to {@link #forwardFrame(Frame, Callback, boolean)}.
+ */
+public abstract class FragmentingFlusher extends TransformingFlusher
+{
+    private static final Logger LOG = Log.getLogger(FragmentingFlusher.class);
+    private final Configuration configuration;
+    private FrameEntry current;
+
+    public FragmentingFlusher(Configuration configuration)
+    {
+        this.configuration = configuration;
+    }
+
+    abstract void forwardFrame(Frame frame, Callback callback, boolean batch);
+
+    @Override
+    protected boolean onFrame(Frame frame, Callback callback, boolean batch)
+    {
+        long maxFrameSize = configuration.getMaxFrameSize();
+        if (frame.isControlFrame() || maxFrameSize <= 0 || frame.getPayloadLength() <= maxFrameSize)
+        {
+            forwardFrame(frame, callback, batch);
+            return true;
+        }
+
+        current = new FrameEntry(frame, callback, batch);
+        boolean finished = fragment(callback, true);
+        if (finished)
+            current = null;
+        return finished;
+    }
+
+    @Override
+    protected boolean transform(Callback callback)
+    {
+        boolean finished = fragment(callback, false);
+        if (finished)
+            current = null;
+        return finished;
+    }
+
+    private boolean fragment(Callback callback, boolean first)
+    {
+        Frame frame = current.frame;
+        ByteBuffer payload = frame.getPayload();
+        int remaining = payload.remaining();
+        long maxFrameSize = configuration.getMaxFrameSize();
+        int fragmentSize = (int)Math.min(remaining, maxFrameSize);
+
+        boolean continuation = (frame.getOpCode() == OpCode.CONTINUATION) || !first;
+        Frame fragment = new Frame(continuation ? OpCode.CONTINUATION : frame.getOpCode());
+        boolean finished = (maxFrameSize <= 0 || remaining <= maxFrameSize);
+        fragment.setFin(frame.isFin() && finished);
+
+        // If we don't need to fragment just forward with original payload.
+        if (finished)
+        {
+            fragment.setPayload(frame.getPayload());
+            forwardFrame(fragment, callback, current.batch);
+            return true;
+        }
+
+        // Slice the fragmented payload from the buffer.
+        int limit = payload.limit();
+        int newLimit = payload.position() + fragmentSize;
+        payload.limit(newLimit);
```

Comment:
I don't think we should be changing the ByteBuffer of the frame to be sent, as I'm pretty sure that sending a frame normally (without extensions) does not modify the payload buffer.

Instead we should be taking a slice at the very beginning and then modifying the slice position and limit as the fragment payload

Reply:
@gregw Just wrote a test for 9.4.x and sending a ByteBuffer of content without any extensions does change the original buffer and move the position to the limit.

Reply:
Well I think that is an issue.  I really don't think it should!

Reply:
I can open a separate issue for that? Doesn't seem like that change should be done in this PR.

Reply:
Sure, but this PR should not make it worse.  Hmmmm or does that make it strange that sometimes we do consume the payload and other times we do not????

Can you point to the code that does the consuming so I can ponder this a bit more?

Reply:
Looks like it is being done in a few places in `FrameFlusher`.

Reply:
`Generator` has a setting for the constructor called `readOnly` which looks like it was intended to not modify the buffer when generating bytes from a frame. It is only ever set as true for the `FrameCaptureExtension`.